### PR TITLE
InstallRepoVersionless

### PR DESCRIPTION
**Pull request by cbridgha**
This PR has changes required to both the install and repo components for supporting versionless features and "platforms" 
1) Changes to InstallUtils that reads the server.xml for a list of features, now returns both features and platforms lists
2) FeatureUtility propogates these values for the "InstallServerAction"
3) Most of these changes I tried to follow the patterns used, buy please let me know if it looks wrong...
4) I added a test, but maybe should be commented out - will only work with Beta flag on...
5) The platforms are now also added to the RepositoryResolver "resolveAsSet" method, and created the old signature that passes a null for platforms...

### CODE REVIEW

**Comment by cbridgha** [comment]
#build

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ExceptionUtils.java
Comment (tbitonti) [inline dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ExceptionUtils.java:1]: Are the System.out's temporary?

File: dev/com.ibm.ws.repository.resolver/src/com/ibm/ws/repository/resolver/RepositoryResolver.java
Comment (tbitonti) [inline dev/com.ibm.ws.repository.resolver/src/com/ibm/ws/repository/resolver/RepositoryResolver.java:1]: More System.out's.

**Review by tbitonti — DISMISSED** [review]
Looks good.  There are System.out's in two of the classes.  I can't tell if those need to be removed.

File: dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ExceptionUtils.java
Comment (cbridgha) [reply dev/com.ibm.ws.install/src/com/ibm/ws/install/internal/ExceptionUtils.java:1]: Thanks I'll remove - missed removing all of them....

**Comment by cbridgha** [comment]
#build

File: dev/com.ibm.ws.repository.resolver/src/com/ibm/ws/repository/resolver/RepositoryResolver.java
Comment (cbridgha) [reply dev/com.ibm.ws.repository.resolver/src/com/ibm/ws/repository/resolver/RepositoryResolver.java:1]: fixed...

File: dev/com.ibm.ws.repository.resolver/src/com/ibm/ws/repository/resolver/RepositoryResolver.java [hunk 1]
```diff
@@ -422,8 +478,21 @@ void initResolve() {
...
+
+    /**
+     * Populates {@link #samplesToInstall} and {@link #featureNamesToResolve} by processing the list of names to resolve and identifying which are samples.
+     *
+     * @param namesToResolve the list of names to resolve
+     * @param platforms
+     */
+    void processNames(Collection<String> namesToResolve, Collection<String> platforms) {
+        if (platforms != null)
+            requestedPlatformNames = platforms;
```
Comment (Azquelt) [inline dev/com.ibm.ws.repository.resolver/src/com/ibm/ws/repository/resolver/RepositoryResolver.java:495]: This doesn't really belong in `processNames`, there's no processing needed here, just set it in `resolve`

**Comment by cbridgha** [comment]
Linking related issues: #28165 and #28321

**Review by Azquelt — COMMENTED** [review]
Needs tests added, I would expect
- unit tests added to `com.ibm.ws.repository.resolver.ResolutionTests`, covering all expected resolution scenarios including any error cases.
- a smaller set of realistic install scenarios added to the install tests

I note that there's no additional information being stored in the ESA or passed through to the kernel resolver. I suspect there are relevant methods in `com.ibm.ws.repository.resolver.internal.kernel.KernelResolverEsa` which need to be implemented and may require additions to the repository metadata. E.g. `getPlatforms`, `isVersionless` - stubs added in #28620. I would also expect tests for these to be added to `KernelResolverEsaTest`.

If there is additional metadata stored in the repository (i.e. new methods on `ESAResource`) then that would also need new tests.

**Review by Azquelt — CHANGES_REQUESTED** [review]
Noting the above comments, I wouldn't expect this to be merged without changes.

**Comment by Azquelt** [comment]
Taking a look at #28321, it looks like the data from the `WLP-Platform` header hasn't been added to the repository metadata. Once it is there, I think it would need to be returned from `KernelResolverEsa.getPlatforms`. I'm not sure how much of the rest of the code should be expected to work without this present.

Update: I think the work required to add the metadata is would be:
- Transport layer: depends on the design of how the information should be represented in the repository, but probably new methods and fields in `WlpInformation`
- Resource layer: new methods on `EsaResource` (getter), `EsaResourceWritable` (setter) and `EsaResourceImpl` (both)
- Parser: update to `EsaParser` to read the header from the manifest
- Appropriate tests for all the above (there should be existing test classes and the additional tests should be small)
  - we may still have internal tests which depend on internal test systems, check with @pnickoll 
- Check that the new information is included in the repo output from the build
  - I think there should be a single-file JSON which would be deployed to maven central and a zip repo which would be offered for download
  - May need to do a personal release build with all the versionless features set to GA if those artifacts aren't generated for beta features
  - I think all these places use `EsaParser` so I'd expect them to work once that's updated, though it's possible that we use a prebuilt jar somewhere, rather than the code that's just been built.
- Check whether any jars used to do the push to the repository need updating.
  - IIRC there may be a jar used in the upload process that needs replaced manually, check with @pnickoll again

File: dev/com.ibm.ws.install.featureUtility_fat/fat/src/com/ibm/ws/install/featureUtility/fat/InstallServerTest.java [hunk 2]
```diff
@@ -155,6 +155,27 @@ public void testInvalidMultiVersionFeatures() throws Exception {
 	checkCommandOutput(po, 21, "CWWKF1405E", null);
 	Log.exiting(c, METHOD_NAME);
     }
+    
+    /**
+     * Test the install of versionless servlet from maven central. Multi-version is not
+     * supported with installServerFeature as it cannot be installed to same
+     * resource.
```
Comment (jjiwooLim) [inline dev/com.ibm.ws.install.featureUtility_fat/fat/src/com/ibm/ws/install/featureUtility/fat/InstallServerTest.java:162]:
```suggestion
     * Test the install of versionless servlet from maven central. "jakartaee-10.0" and "servlet" configuration should install servlet-6.0.
```

File: dev/com.ibm.ws.install.featureUtility_fat/fat/src/com/ibm/ws/install/featureUtility/fat/InstallServerTest.java [hunk 3]
```diff
@@ -155,6 +155,27 @@ public void testInvalidMultiVersionFeatures() throws Exception {
...
+    public void testVersionlessWithPlatformFeatures() throws Exception {
+	final String METHOD_NAME = "testInvalidMultiVersionFeatures";
+	Log.entering(c, METHOD_NAME);
+
+	copyFileToMinifiedRoot("usr/servers/serverX", "publish/tmp/versionlessWPlatform/server.xml");
+
+	String[] param1s = { "installServerFeatures", "serverX", "--verbose" };
+	ProgramOutput po = runFeatureUtility(METHOD_NAME, param1s);
+
+	checkCommandOutput(po, 21, "CWWKF1405E", null);
```
Comment (jjiwooLim) [inline dev/com.ibm.ws.install.featureUtility_fat/fat/src/com/ibm/ws/install/featureUtility/fat/InstallServerTest.java:176]:
```suggestion
 String[] filesList = { "/lib/features/com.ibm.websphere.appserver.servlet-6.0.mf" };
	checkCommandOutput(po, 0, null, filesList);
```

**Review by jjiwooLim — CHANGES_REQUESTED** [review]
Install implementation looks good! I suggested some changes for the test.  Btw, should featureUtility read `PREFERRED_PLATFORM_VERSIONS` environment variable and pass it to the resolver?

**Comment by cbridgha** [comment]
@Azquelt - thank you for your comments, and first of all I didn't mean this to be merged - this was my experiment to see what changes were required to get some function working.....    There are many questions to answer - and I will try to setup a call next week.  I do question if the ESA needs to be changed to capture the new WLP-Platform Manifest header, but I also realize any API that may be needed during resolution needs to be provided by us.  the wlp-platform value is only used in a preprocessing step that allows some optimization before performing the actual resolution - but more to come here...  also working on the API docs you mentioned...

**Comment by cbridgha** [comment]
@jjiwooLim thanks for taking a look, and as I mentioned to @Azquelt  - this was an experiment, not meant to be merged... but hopefully helpful in the changes that will be required.   Good question on the env variable, and yes we need to provide an api that helps pass this value (we do this internally, but will expose it as an spi...   ultimately we need to expose an API that determines a "target platform" that makes sense with the feature set specified, and rather than duplicating implmentation - we'll provide. - I'll setup a call late next week that explains some of these details.

**Comment by cbridgha** [comment]
Closing this - Was used for prototyping purposes....

**Pull request closed without merging**